### PR TITLE
refactor: refactor merge into optimizer and fix unexpected distribution plan

### PR DESCRIPTION
**Pull request by xudong963**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

1. support explain pipeline merge into
2. support explain fragments merge into
3. reduce the occurrence of `*distributed && !*change_join_order`
4. inner join can directly use broadcast join, don't need to add `RowNumber` operator
5. reduce the complexity of merge into optimizer.
6. improve merge into distribution plan, see the file `tests/sqllogictests/suites/mode/cluster/merge_into_non_equal_distributed.test`

- Fixes #[Link the issue here]

## Tests

- [ ] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test  - _Explain why_

## Type of change

- [x] Bug Fix (non-breaking change which fixes an issue)
- [ ] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [x] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/15507)

### CODE REVIEW

File: src/query/storages/fuse/src/operations/merge_into/mutator/matched_mutator.rs [hunk 1]
```diff
@@ -479,15 +486,3 @@ impl AggregationContext {
         Ok(Some(mutation))
     }
 }
-
-pub(crate) fn get_row_id(data_block: &DataBlock, row_id_idx: usize) -> Result<Buffer<u64>> {
```
Comment (xudong963) [inline src/query/storages/fuse/src/operations/merge_into/mutator/matched_mutator.rs:483]: The method doesn't consider `Scalar`

File: src/query/service/src/interpreters/interpreter_merge_into.rs [hunk 2]
```diff
@@ -74,8 +72,8 @@ pub struct MergeIntoInterpreter {
 }
 
 impl MergeIntoInterpreter {
-    pub fn try_create(ctx: Arc<QueryContext>, plan: MergePlan) -> Result<InterpreterPtr> {
```
Comment (Dousir9) [inline src/query/service/src/interpreters/interpreter_merge_into.rs:77]: Keep Result<InterpreterPtr>

File: src/query/sql/src/planner/optimizer/optimizer.rs [hunk 3]
```diff
@@ -502,115 +469,35 @@ async fn optimize_merge_into(opt_ctx: OptimizerContext, plan: Box<MergeInto>) ->
...
-        // III otherwise, use `merge_into_join_sexpr` as standalone execution(so if change join order is false,but doesn't match_pattern, we don't support distributed,in fact. case I
-        // can take this at most time, if that's a hash shuffle, the I can take it. We think source is always very small).
-        // input is a Join_SExpr
-        let mut merge_into_join_sexpr =
-            optimize_distributed_query(opt_ctx.table_ctx.clone(), &join_sexpr)?;
+    if opt_ctx.enable_distributed_optimization {
         let merge_source_optimizer = MergeSourceOptimizer::create();
-        // II.
-        // - join spilling is disabled
-        let (optimized_distributed_merge_into_join_sexpr, distributed) = if opt_ctx
```
Comment (Dousir9) [inline src/query/sql/src/planner/optimizer/optimizer.rs:527]: Is distributed supported in case of hash spill ?

Comment (xudong963) [reply src/query/sql/src/planner/optimizer/optimizer.rs:527]: use hash join spill

File: src/query/service/src/interpreters/interpreter_merge_into.rs (hunk 2 above)
Comment (xudong963) [reply src/query/service/src/interpreters/interpreter_merge_into.rs:77]: Align with other interpreter

Comment (Dousir9) [reply src/query/service/src/interpreters/interpreter_merge_into.rs:77]: `Plan::Insert` and `Plan::Replace` are not aligned.

**Review by Dousir9 — APPROVED** [review]

File: src/query/service/src/interpreters/interpreter_merge_into.rs (hunk 2 above)
Comment (xudong963) [reply src/query/service/src/interpreters/interpreter_merge_into.rs:77]: let's refactor them later

**Pull request merged**